### ISSUE TITLE
Why ROCm applications directly call hsa-runtime APIs since 2.3.x?

### ISSUE THREAD

**Issue by ghostplant** [ROCm-Developer-Tools/HIP#1047]
This is the output of ROCm application compiled with rocm-2.2:
```sh
/root$ nm -D ./rocm_sample.out | grep -e "U hsa" -e "U hip"
                 U hipInit
                 U hipMalloc
                 U hipMemcpy
                 U hipSetDevice
```

This is the output of ROCm application compiled with rocm-2.3:
```sh
/root$ nm -D rocm_sample.out | grep -e "U hsa" -e "U hip"
                 U hipInit
                 U hipMalloc
                 U hipMemcpy
                 U hipModuleLaunchKernel
                 U hipSetDevice
                 U hsa_agent_get_info
                 U hsa_agent_iterate_isas
                 U hsa_amd_memory_lock
                 U hsa_code_object_reader_create_from_memory
                 U hsa_code_object_reader_destroy
                 U hsa_executable_agent_global_variable_define
                 U hsa_executable_create_alt
                 U hsa_executable_freeze
                 U hsa_executable_iterate_agent_symbols
                 U hsa_executable_load_agent_code_object
                 U hsa_executable_symbol_get_info
                 U hsa_isa_from_name
                 U hsa_system_get_major_extension_table
                 U hsa_system_major_extension_supported
```

I am not sure the purpose of such changes, but wonder whether it is possible to avoid such changes in the future release?

**Comment by mangupta** [comment]
This is an intended change which was introduced via PR #929. The PR also describes the rationale for this change.
I am curious to know how are you impacted by having these additional symbols exposed?

**Comment by ghostplant** [comment]
@mangupta Yeah, a big impact to make ROCm support windows platform..

**Comment by ghostplant** [comment]
@mangupta I think for the new method, it spends a lot more memory, right?
For example, assume a running program contains kernel code A, B and C, when it loads another library with kernel code D, another **code_object_blobs dictionary** will load all kernels A, B, C, and D, while the original **code_object_blobs dictionary** is not freed. Thus the code of A, B and C are meaningless duplicated many times. So why a global one was not recommended?

**Comment by gargrahul** [comment]
@ghostplant I believe your concern is being addressed by #1057.

**Comment by ghostplant** [comment]
@gargrahul Thanks, but I think it is a little weird that after #929 and #1047, there are actually no much difference with it originally was in ROCm 2.2, as the original algorithm is when a kernel code is not detecting (in other word, hit cache), the database would be reloaded to keep latest.

The most critical thing regardless of memory duplication is that library-link pollution problem which doesn't exist in ROCm 2.2, and both #1052 and https://github.com/scchan/HIP/tree/lazy_binding-merged_program_state are trying to survive this problem.
